### PR TITLE
feat(devx): Local development with devcontainers

### PR DESCRIPTION
**Pull request by AndesKrrrrrrrrrrr**
Add Devcontainer support for local development

## Testing
1. Install the Devcontainers VSCode extension
2. Click "re-open in devcontainer" in lower-right corner (takes some time)
3. Try to do normal dev stuff

## Summary by CodeRabbit

- **New Features**
  - Introduced support for development containers with Docker, including multi-stage build processes and automated setups for VS Code.
  - Added AWS configuration profiles for different environments (dev, staging, prod) with SSO configurations and region settings.
  - Included new development-related scripts in `package.json`.

- **Documentation**
  - Updated `README.md` with detailed instructions on setting up and using development containers.
  - Updated `scripts/README.md` to reference `xdg-open` instead of "Firefox" for better compatibility.

### CODE REVIEW

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  java-11-openjdk g++ jq yq
+# Set up AWS specific env
+RUN --mount=type=cache,target=/var/cache/dnf \
+  dnf install -y \
+  awscli2
+
+# Niceties
+RUN --mount=type=cache,target=/var/cache/dnf \
+  dnf install -y \
+  ${SHELL} the_silver_searcher ripgrep fzf neovim
```
[the remark this answers is not in the document]
Comment (AndesKrrrrrrrrrrr) [reply .devcontainer/Dockerfile:35]: This doesn't make sense. If we need to change the package manager, we'll have to change the install command and possibly package names as well. The entire `RUN` command would have to be rewritten

**Review by saevarma — COMMENTED** [review]
What is the usage for this?

This is a docker container that runs ubuntu with installed node and shell  🤷‍♂️

File: .devcontainer/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+# Build dependencies
+RUN --mount=type=cache,target=/var/cache/apt \
+  apt-get update && \
+  apt-get install -y --no-install-recommends \
+  make cmake g++
+
+# Install cargo stuff
+RUN --mount=type=cache,target=/root/.cargo \
+  cargo install starship
+
```
[the remark this answers is not in the document]
Comment (AndesKrrrrrrrrrrr) [reply .devcontainer/Dockerfile:12]: There is only 1 cargo install instruction, which even is in its own output, further reducing cache busting

**Comment by AndesKrrrrrrrrrrr** [comment]
> This is a docker container that runs ubuntu with installed node and shell 🤷‍♂️

Yes, that is the point. https://containers.dev/

This should simplify getting started and lessen our support load. Never again:
- "Are you sure you've installed the correct version of Java?"
- "Are you using another shell than `bash`?"
- "Ah, you have to set the right `node` version, see these docs..." :smile:

File: .devcontainer/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,64 @@
+FROM rust:latest as cargo
+
+# Build dependencies
+RUN --mount=type=cache,target=/var/cache/apt \
+  apt-get update && \
+  apt-get install -y --no-install-recommends \
+  make cmake g++
```
[the remark this answers is not in the document]
Comment (AndesKrrrrrrrrrrr) [reply .devcontainer/Dockerfile:7]: How does specifying a specific version reduce image size?

[the remark this answers is not in the document]
Comment (AndesKrrrrrrrrrrr) [reply .devcontainer/Dockerfile]: Pinned, just for you :heart:

**Comment by robertaandersen** [comment]
> What is the usage for this?
> 
> This is a docker container that runs ubuntu with installed node and shell 🤷‍♂️

Also this saves a lot of onboarding time as one could in theory define all vscode extensions deemed necessary for developers to have a happy day :)

File: .devcontainer/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  cargo install starship
+
+FROM docker.io/rust:${RUST_VERSION} as cargo-volta
+WORKDIR /build/volta
+RUN --mount=type=cache,target=/root/.cargo/ --mount=type=cache,target=/root/.cache/ \
+  git init && \
+  (git remote | grep -q origin || git remote add origin https://github.com/volta-cli/volta) && \
+  git fetch origin && \
+  git checkout origin/main --force && \
+  cargo build --release
```
[the remark this answers is not in the document]
Comment (AndesKrrrrrrrrrrr) [reply .devcontainer/Dockerfile:22]: Good suggestion. Maybe at another time

**Comment by saevarma** [comment]
> > This is a docker container that runs ubuntu with installed node and shell 🤷‍♂️
> 
> Yes, that is the point. https://containers.dev/
> 
> This should simplify getting started and lessen our support load. Never again:
> 
> * "Are you sure you've installed the correct version of Java?"
> * "Are you using another shell than `bash`?"
> * "Ah, you have to set the right `node` version, see these docs..." 😄

Is there any documentation for devs on how to use this?

**Comment by AndesKrrrrrrrrrrr** [comment]
> Is there any documentation for devs on how to use this?

Good point. I'll add some documentation to the `README.md`

File: .devcontainer/linux/devcontainer.json [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+  "build": { "dockerfile": "Dockerfile", "context": "../" },
+  "customizations": {
+    "vscode": {
+      "extensions": ["dbaeumer.vscode-eslint"]
+    }
+  },
+  "runArgs": ["--userns=keep-id"],
+
+  "forwardPorts": [3000]
+}
```
Comment (AndesKrrrrrrrrrrr) [inline .devcontainer/linux/devcontainer.json:12]: @coderabbitai write documentation on how developers utilize devcontainers in the root README

Comment (AndesKrrrrrrrrrrr) [reply .devcontainer/linux/devcontainer.json:12]: Please make this a code suggestion for easy inclusion

**Review by saevarma — APPROVED** [review]
🚀

**Review by svanaeinars — APPROVED** [review]
LGTM

File: .devcontainer/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+# Tried pinning versions, but got errors from version solver (DL3041).
+# hadolint ignore=DL3040,DL3041
+RUN --mount=type=cache,target=/var/cache/dnf \
+  dnf install -y \
+  # Basic requirements
+  bash zsh git java-11-openjdk gcc-c++ jq graphviz \
+  # AWS/DevOps specific dependencies
+  awscli2 \
+  # Niceties
+  the_silver_searcher ripgrep fzf neovim
```
[the remark this answers is not in the document]
Comment (AndesKrrrrrrrrrrr) [reply .devcontainer/Dockerfile]: Not needed, since we're using cache mounts

**Review by saevarma — APPROVED** [review]

**Pull request merged**